tor_ocr/main.py

```diff
@@ -42,28 +42,49 @@
 """
 
 
-def process_image(local_file):
-    with PyTessBaseAPI() as api:
-        api.SetImageFile(local_file)
-        text = api.GetUTF8Text()
+def process_image(image_url):
+    # If you feed it a regular image with no text, more often than not
+    # you'll get newlines and spaces back. We strip those out to see if
+    # we actually got anything of substance.
 
-        confidences = api.AllWordConfidences()
-        if not confidences or len(confidences) == 0:
-            # we have an image, but it *really* couldn't find anything, not
-            # even false positives.
-            return None
+    json_result = json.loads(ocr_space_url(image_url))
 
-        logging.debug('Average of confidences: {}'.format(
-            sum(confidences) / len(confidences))
+    result = {
+        'text': json_result['ParsedText'],
+        'exit_code': json_result['OCRExitCode'],
+        'error_on_processing': json_result['IsErroredOnProcessing'],
+        'error_message': json_result['ErrorMessage'],
+        'error_details': json_result['ErrorDetails'],
+        'process_time_in_ms': json_result['ProcessingTimeInMilliseconds'],
+    }
+
+    error_codes = {
+        1: 'success',
+        0: 'file not found',
+        -10: 'OCR engine parse error',
+        -20: 'timeout',
+        -30: 'validation error',
+        -99: 'UNKNOWN ERROR',
+    }
+
+    if result['text'].strip() == '':
+        return None
+
+    if result['exit_code'] != 1 or result['error_on_processing']:
+        logging.warning(
+            'OCR Error encountered: ' +
+            error_codes[result['exit_code']] +
+            '. Error message: ' +
+            result['error_message'] +
+            '. Error details: ' +
+            result['error_details'] +
+            '.'
```

Comment:
This is a really great time to raise a custom exception.

```python
class OCRError(Exception):
    CODES = {
        # [move the map of codes to human readable meanings here]
    }

    def __init__(self, result):
        super(self.__class__, self).__init__(self.CODES[result['exit_code']])
        self.result = result
```

Then we raise it like this in order to handle it higher up the stack:

```python
if result['exit_code'] != 1 or result['error_on_processing']:
    raise OCRError(result)
```

The scope of this method is to do the work and provide a consistent interface with a third-party resource. We've done that. Now we need to continue with that consistent interface in a way we can handle with other methods, meaning raising an exception if an error occurs.

Thoughts?

Reply:
Seems like a lot of boiler plate for a one time use. Also, we don't want to always throw an exception, i.e. if the request times out, or no text is found.

Reply:
We may not want to throw an exception all the time, but throwing an exception is great so that we can catch and filter if necessary. Also, an exception will get sent to Bugsnag (and, in turn, Slack) so that we're more aware of it happening :)

Reply:
Also of note, the timed out response or other breakdown at the HTTP level between our bot and the API will already throw an exception, courtesy of the requests library.

If you mean a timeout between the API and the image hoster, we can add retry logic in later. Since we’ll be passing the API response data with the exception we have a lot of flexibility on if and when we might want to retry

Reply:
Yeah, the API can timeout while doing the OCR. Retry logic wouldn't be a good idea, because a timeout = too much for the API on the free plan, at least from what I can gather from the docs.

Also, in that case I'll add the exception.